### ISSUE TITLE
Deadlock when calling threadpool_limits from python threadpool

### ISSUE THREAD

**Issue by jeremiedbb** [joblib/threadpoolctl#40]
In very specific conditions.

To reproduce: 
- environment with numpy, mkl
- clang compiler

`CC_OUTER_LOOP=clang continuous_integration/build_test_ext.sh`
(inner loop compiler is irrelevant)

```python
from threadpoolctl import threadpool_info 
import numpy as np 
from tests._openmp_test_helper.nested_prange_blas import check_nested_prange_blas 
from multiprocessing.pool import ThreadPool 
 
A = np.ones((1000, 10)) 
B = np.ones((100, 10)) 
 
# blas call within a prange
# remove this and no more deadlock
# Lower number of threads (4) and no more deadlock
check_nested_prange_blas(A, B, 4)

# Remove this and no more deadlock.
# Put it before check_nested_prange_blas and no more deadlock
np.dot(A, B.T) 
 
def func(i): 
    threadpool_info()
    # threadpool_info() can be replaced by
    # check_nested_prange_blas(A, B, 1)
    # the 1 is important
     
def test(): 
    pool = ThreadPool(2) 
    pool.map(func, range(2)) 
    pool.close() 
    pool.join()  # no join = no more deadlock
 
test() 
test()  # remove the second one and no more deadlock
```

@ogrisel @tomMoral

**Comment by jeremiedbb** [comment]
Additional informations

- **libiomp** raises the following warning during the call to `check_nested_prange_blas`:
```
OMP: Warning #190: Forking a process while a parallel region is active is potentially unsafe.
```

- It's the call to `get_num_threads` of **libomp** in threadpool_info which causes the deadlock

**Comment by ogrisel** [comment]
> libiomp raises the following warning during the call to check_nested_prange_blas:
> OMP: Warning #190: Forking a process while a parallel region is active is potentially unsafe.

This is unexpected no? This code does not seem to do any fork.

**Comment by jeremiedbb** [comment]
`check_nested_parange_blas` is essentially:
```
for i in prange(n):
    blas_gemm(...)
```

If I interpret correctly the warning, the "activve parallel region" is the outer OpenMP (libomp) parallel region. And the "fork" is libiomp creating new threads (OpenMP thread creation is through fork, see https://en.wikipedia.org/wiki/OpenMP)

**Comment by ogrisel** [comment]
Ah ok but this is not equivalent to a POSIX fork system call. Just ambiguous naming. So for some reason intel-openmp used by BLAS can detect that it's being called inside a clang/libomp parallel region.

**Comment by jeremiedbb** [comment]
Are you sure ? because on the wikipedia page, the description of the design links to https://en.wikipedia.org/wiki/Fork_(system_call)

**Comment by jeremiedbb** [comment]
Notice that this situation is not exactly our use case. When calling  
`check_nested_prange_blas(A, B, 4)`  
you ask 4 threads for the prange and let mkl use as many as possible

Our use case would be to add:  
```
with threadpool_limits(1, 'blas'):
    check_nest_prange_blas(A, B, 4)
```
In that case there's no deadlock anymore.

**Comment by ogrisel** [comment]
Linux threads are spawned by a clone system call: https://en.wikipedia.org/wiki/Fork_(system_call)#Clone

**Comment by ogrisel** [comment]
So apparently it does not seem to be possible to safely dlopen both llvm-openmp and intel-openmp in the same Python program under Linux (it works without any problem under macOS apparently...).

In the Python ecosystem, the main reason to have a program link against intel-openmp would be to use the numpy and scipy packages from anaconda that use MKL by default. Fortunatelt MKL can be reconfigured at runtime by setting the `MKL_THREADING_LAYER=GNU` or `MKL_THREADING_LAYER=TBB` to use either libgomp or Threading Building Block instead of intel-openmp (libiomp).

So the plan of action would be to have threadpoolctl warn the user whenever intel-opemp and llvm-openmp are detected both at the same time under Linux and let the user know about possible workarounds:

- set the environment variable `MKL_THREADING_LAYER=GNU` to let numpy / scipy use 
- `conda install tbb` and then `MKL_THREADING_LAYER=TBB`
- `conda install nomkl` to use openblas instead of MKL
- recompile compiled extensions that use OpenMP with gcc/libgomp instead of clang/llvm

**Comment by jeremiedbb** [comment]
Kind of fixed in #49 (nothing we can do about that. Just warn the user)
